miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt

```diff
@@ -1,30 +1,60 @@
 package com.rakuten.tech.mobile.miniapp.js
 
+import android.app.Activity
+import android.content.Context
 import android.webkit.JavascriptInterface
+import androidx.annotation.VisibleForTesting
+import androidx.core.app.ActivityCompat
 import com.google.gson.Gson
 import com.rakuten.tech.mobile.miniapp.display.WebViewListener
 
+@VisibleForTesting
+internal const val GET_UNIQUE_ID = "getUniqueId"
+@VisibleForTesting
+internal const val REQUEST_PERMISSION = "requestPermission"
+
 /** Bridge interface for communicating with mini app. **/
-abstract class MiniAppMessageBridge {
+abstract class MiniAppMessageBridge(val context: Context) {
     private lateinit var webViewListener: WebViewListener
 
     /** Get provided id of mini app for any purpose. **/
     abstract fun getUniqueId(): String
 
     /** Handle the message from external. **/
-    @Suppress("TooGenericExceptionCaught", "SwallowedException")
+    @Suppress("TooGenericExceptionCaught", "SwallowedException", "LongMethod")
     @JavascriptInterface
     fun postMessage(jsonStr: String) {
         val callbackObj = Gson().fromJson(jsonStr, CallbackObj::class.java)
+
         when (callbackObj.action) {
-            "getUniqueId" -> try {
+            GET_UNIQUE_ID -> try {
                 postValue(callbackObj.id, getUniqueId())
             } catch (e: Exception) {
                 postError(callbackObj.id, "Cannot get unique id: ${e.message}")
             }
+            REQUEST_PERMISSION -> try {
+                val permissionParam = Gson().fromJson(callbackObj.param, Permission::class.java)
+                requestPermission(
+                    MiniAppPermission.getPermissionRequest(permissionParam.permission),
+                    MiniAppPermission.getRequestCode(permissionParam.permission))
+            } catch (e: Exception) {
+                postError(callbackObj.id, "Cannot request permission: ${e.message}")
+            }
         }
     }
 
+    /** Inform the permission request result to MiniApp. **/
+    fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
+        if (permissions.size == grantResults.size) {
+            for (i in permissions.indices)
+                webViewListener.onRequestPermissionsResult(requestCode, permissions[i], grantResults[i])
+        }
+    }
+
+    /** Post permission request from external. **/
+    internal fun requestPermission(permissions: Array<String>, requestCode: Int) =
+        ActivityCompat.requestPermissions(context as Activity, permissions, requestCode)
```

Comment:
Sorry, I think the ticket for this was created before we had talked about how we want to handle permissions.

I think that we should let the Host App handle the request for device permissions. So we can add a method to `MiniAppMessageBridge` for something like `onRequestPermission` and then just pass the permission to the Host App as an enum or something. Does that sound okay @minh-rakuten @ts-nitish-a-jain?

Also, we need to be able to grant the permission to each Mini App independently of each other and cache which Mini Apps have been granted each permission. So the Host App would potentially show two permissions dialogs: the native permissions dialog and then also a custom dialog asking to grant permission to that particular Mini App. However, I think this task was only for getting the location permission working, so we can wait to implement the caching and independent permissions at a later time.

Reply:
Device Permission access per mini-app is interesting. This would need confirmation on the product side if they wish to show the next popup or grant implicitly.
From a secure zone pov, it makes sense to have two dialogs. 
At the same time, caching can result in a bloated structure for maintaining the intelligence of permission access per mini-app.

Reply:
@ts-nitish-a-jain Device permission per Mini App was already discussed by the product side - I think this is what's expected. We had discussed using similar mechanisms as the custom permissions for caching which permissions have been granted.

In the end, it's really up to the Host App though. If they don't care about granting it on a per Mini App basis, then they don't have to show a Mini App specific permissions dialog and can just implicitly grant the permission to all Mini Apps.